engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceCreationCreatedApplier.java

```diff
@@ -0,0 +1,107 @@
+import io.camunda.zeebe.engine.state.immutable.ProcessState;
+import io.camunda.zeebe.engine.state.instance.ElementInstance;
+import io.camunda.zeebe.engine.state.mutable.MutableElementInstanceState;
+import io.camunda.zeebe.engine.state.mutable.MutableProcessState;
+import io.camunda.zeebe.protocol.impl.record.value.processinstance.ProcessInstanceCreationRecord;
+import io.camunda.zeebe.protocol.record.intent.ProcessInstanceCreationIntent;
+import io.camunda.zeebe.protocol.record.value.BpmnElementType;
+import java.util.List;
+import org.agrona.DirectBuffer;
+
+final class ProcessInstanceCreationCreatedApplier
+    implements TypedEventApplier<ProcessInstanceCreationIntent, ProcessInstanceCreationRecord> {
+
+  private final ProcessState processState;
+  private final MutableElementInstanceState elementInstanceState;
+
+  public ProcessInstanceCreationCreatedApplier(
+      final MutableProcessState processState,
+      final MutableElementInstanceState elementInstanceState) {
+    this.processState = processState;
+    this.elementInstanceState = elementInstanceState;
+  }
+
+  @Override
+  public void applyState(final long key, final ProcessInstanceCreationRecord value) {
+    if (value.hasStartInstructions()) {
+      final var process =
+          processState.getProcessByKey(value.getProcessDefinitionKey()).getProcess();
+      final ElementInstance processInstance =
+          elementInstanceState.getInstance(value.getProcessInstanceKey());
+
+      value.getStartInstructions().stream()
+          .map(instruction -> process.getElementById(instruction.getElementId()))
+          .filter(element -> element.getElementType().equals(BpmnElementType.PARALLEL_GATEWAY))
+          .map(ExecutableFlowNode.class::cast)
+          .forEach(
+              element -> {
+                final var parentElementId = element.getFlowScope().getId();
+                final ElementInstance flowScope =
+                    findParentFlowScope(processInstance, parentElementId);
+                incrementNumberOfTakenSequenceFlows(element, flowScope);
+              });
+    }
+  }
+
+  /**
+   * Traverses the element instances to find one that matches the parent element id. If this is the
+   * process instance, it is returned immediately. This will work because we will always activate
+   * the flow scope of a start instruction once.
+   *
+   * @param processInstance the highest element instance of a process
+   * @param targetElementId the id we are looking for
+   * @return the element instance which matches the targetElementId
+   */
+  private ElementInstance findParentFlowScope(
+      final ElementInstance processInstance, final DirectBuffer targetElementId) {
+    final ElementInstance flowScope;
+    if (processInstance.getValue().getElementIdBuffer().equals(targetElementId)) {
+      flowScope = processInstance;
+    } else {
+      flowScope = findFlowScopeInChildren(processInstance, targetElementId);
+    }
+    return flowScope;
```

Comment:
🔧 This is just nitpicking, but there's no need for single return in Java code. Early returns are generally favored instead.

```java
if (processInstance.getValue().getElementIdBuffer().equals(targetElementId)) {
  return processInstance;
}

return findFlowScopeInChildren(processInstance, targetElementId);
```